src/petals/server/backend.py

```diff
@@ -91,6 +91,8 @@ def get_inference_cache_descriptors(self, batch_size: int, max_length: int) -> S
         cache_tensors = []
         for device, num_heads in zip(self.module.devices, self.shard_num_heads):
             num_heads //= self.config.num_key_value_groups
+            if hasattr(self.config, "num_key_value_heads"):
+                num_heads = self.config.num_key_value_heads
```

Comment:
Does this still work when the layer is sharded?

Reply:
This works if layer is sharded, unless the user manually rewrites petals to specify an incompatible sharding config into TensorParallel. Without that, TensorParallel will use a default (automatic) config where
- only linear layers are sharded
- each shard computes the full attention kernel, relu, etc


If user shards the Mixtral model, it will be using default config and run  @artek0chumak  code correctly. Because each layer shard requires full KV cache and has full number of heads in config.


In other words, the current code will work in all cases except when user goes out of their way to modify TensorParallel config in a way that breaks petals servers, which would require changing run_server code.